### ISSUE TITLE
[pcl_filters, pcl_surface, ...] Access violation in VTK on dynamic library unloading.

### ISSUE THREAD

**Issue by MK-3PP** [PointCloudLibrary/pcl#4716]
When certain PCL modules are loaded and unloaded dynamically under Windows, there is an [Access Violation in VTK](https://gitlab.kitware.com/vtk/vtk/-/issues/18180).

The issue was initially observed in VTK 8.2 binaries that shipped with PCL 1.11.0. It was then confirmed in own builds of VTK 8.2.0 and VTK 9.0.1. It is passed on to all PCL versions using at least these VTK versions.
VTK 8.2.0, VTK 9.0.1 and test program built with VC16.

This triggers the issue:
```
#include <Windows.h>

int main() {
    auto    cvtk = AddDllDirectory(L"D:\\PCL 1.11.0\\3rdParty\\VTK\\bin");
    auto    cpcl = AddDllDirectory(L"D:\\PCL 1.11.0\\bin");
    HMODULE hMod = LoadLibraryEx(L"pcl_surfaced.dll", NULL, LOAD_LIBRARY_SEARCH_USER_DIRS);

    if (hMod) {
        BOOL freed = FreeLibrary(hMod); // Access Violation in vtkCommonCore-8.2-gd.dll.
    }

    return 0;
}
```

Older versions of PCL (tested in 1.8.1) that came without PCL have no such issue when dynamically loading and unloading.

_We are using prebuilt MSVC2019 releases of PCL. PCL is headless in our application, but modules such as pcl_filters and pcl_surface depend on VTK. This hughely increases the application's footprint, because it requires us to ship (and load) all VTK modules as well._

**Comment by larshg** [comment]
I cannot reproduce this, when the exe is compiled in debug and loads either debug or release version of 
`vtkFiltersGeneral-9.0d.dll`.

However, if I have the exe compiled in release mode and load the debug version of `vtkFiltersGeneral-9.0d.dll` it gives the same access violation.

So could it be a mix of debug / release ?

**Comment by MK-3PP** [comment]
No, I explicitly have tested exe and library in same mode. The issue is present on matching modes on my machine. Mixed modes do fail on LoadLibrary (hMod is NULL) with my setup due to other missing libraties (CRT?).
The exe project is a fresh C++ Console Application. Platform is always amd64.

But I do remember having had mixed modes in earlier, less structured tests where I put a bunch of DLLs in one folder. I saw the same behaviour there. But thanks for pointing that out. If I get to work on this again, starting tomorrow, I will work towards checking matching and mixed modes.

Also, the AddDllDirectory was inserted to allow for fast switching of paths without having to copy tens of megabytes of files. As you can see, i tested a bunch of versions / configs. Actually more than I wrote, but with a less structured approach.

**Comment by larshg** [comment]
I just tried to create a empty console c++ project in VS2019 - and here I can't get it to load anything. But out time for now.

**Comment by MK-3PP** [comment]
**TL;DR:**
- **On my End, the issue is not triggered with exe/lib in mixed config (Dbg/Rel, Rel/Dgb), because `LoadLibrary `fails early.**
- **On my End, the issue is triggered with exe/lib in matching config (Rel/Rel, Dbg/Dbg).**


If I mix modes with a fresh setup (exe is Release, library is Debug configuration), dll loading stops at `vtkCommonCore-8.2-gd.dll` and then everything gets unloaded, before the call to `LoadLibrary` returns.

Trace:
```
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_surfaced.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\rpcrt4.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\user32.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\win32u.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\gdi32.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\gdi32full.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\msvcp_win.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\advapi32.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_octreed.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\msvcrt.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_kdtreed.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_searchd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\sechost.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_commond.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonExecutionModel-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkFiltersCore-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkFiltersModeling-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonDataModel-8.2-gd.dll" loaded. 
**"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonCore-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_searchd.dll" was unloaded.**
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_kdtreed.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_octreed.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_commond.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkFiltersModeling-8.2-gd.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkFiltersCore-8.2-gd.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonExecutionModel-8.2-gd.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonDataModel-8.2-gd.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonCore-8.2-gd.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\win32u.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\msvcp_win.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\user32.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\gdi32.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\gdi32full.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\msvcrt.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\sechost.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\rpcrt4.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\advapi32.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_surfaced.dll" was unloaded.
```
("geladen" -> "loaded", "wurde entladen" -> "was unloaded" manually translated from german.)

I do not know why Loading worked with mixed modes before. Or why it does for you. Maybe it has to do with copying all DLLs in one folder, since I remember doing so, too when I had my early tests.

I thought CRT was the cause. But if I start both, exe and library, in Debug config, it succeeds and i see msvcrt.dll getting loaded and not msvcrtd.dll, as I expected. Also note the double load and then one unload of some of the VTK files, vtkCommonSystem-8.2-gd.dll for example:

```
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_surfaced.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\rpcrt4.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_searchd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_octreed.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_kdtreed.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\user32.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\win32u.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\gdi32.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\gdi32full.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkFiltersModeling-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\bin\pcl_commond.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\msvcp_win.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkFiltersCore-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\ucrtbase.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\advapi32.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\msvcrt.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonDataModel-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\sechost.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonExecutionModel-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonCore-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonTransforms-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonMisc-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkFiltersSources-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkFiltersGeneral-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonMath-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonSystem-8.2-gd.dll" loaded. 
**"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonSystem-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonSystem-8.2-gd.dll" was unloaded.**
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\ws2_32.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtksys-8.2-gd.dll" loaded. 
**"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtksys-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtksys-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtksys-8.2-gd.dll" was unloaded.
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtksys-8.2-gd.dll" was unloaded.**
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\psapi.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\shell32.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\cfgmgr32.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\SHCore.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\combase.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\bcryptprimitives.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\windows.storage.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\profapi.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\powrprof.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\umpdc.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\shlwapi.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\kernel.appcore.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\cryptsp.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "D:\PCL 1.11.0\3rdParty\VTK\bin\vtkCommonComputationalGeometry-8.2-gd.dll" loaded. 
"PCL FreeLibrary.exe" (Win32): "C:\Windows\System32\imm32.dll" loaded. 
```

But I might be hunting spectres here.

**Comment by MK-3PP** [comment]
Update: I have again tested mixed exe/libraries configurations and did not get them to successfully load using PCL DLLs as an entry point. This is contrary to the findings of @larshg.

**Comment by larshg** [comment]
After I created a clean c++ console app I can reproduce the error. I had to add  `| LOAD_LIBRARY_SEARCH_DEFAULT_DIRS` to the `LoadLibraryEx` for some whatever dependencies.

But it surely fail at unloading some VTK libraries.

However I don't think its a PCL issue, so probably can't do much about it, here.

**Comment by MK-3PP** [comment]
I agree about the location of the issue in VTK, hence [my ticket there](https://gitlab.kitware.com/vtk/vtk/-/issues/18180).
But wanted to raise awareness that PCL pulled in an issue via a new dependency to VTK. And more so, while VTK is solely used for Visualization(?), the dependency is loaded even if no visualization takes place.

**It makes us uneasy, having to include 246 new VTK libraries with a cumulative size of 218 MB in PCL 1.11.0 to our application dependencies AND catching a new bug, leaving our dynamic plugin loading mechanism broken.**

**Comment by larshg** [comment]
> But wanted to raise awareness that PCL pulled in an issue via a new dependency to VTK

As far as I know, the dependency for surface module has been there for quite a long time. However, its optional as long as you don't need to use `vtk_smoothing` (skimming the CMakeLists).

So you could compile PCL yourself without VTK installed?

**Comment by MK-3PP** [comment]
I fear this might be necessary. CMake projects under Windows are very painful, especially, when you have to check the resulting libraries into version control.

Last version I had configured and built was 1.8.1. I thought it came without VTK, but as it seems i just disabled it and excluded any reference to it from version control. Thanks for pointing that out.

**Comment by larshg** [comment]
Closing this, since its not a bug in PCL per se and VTK has closed it as well.

Hope you can manage by building PCL yourself, with VTK disabled.

**Comment by kunaltyagi** [comment]
> PCL is headless in our application, but modules such as pcl_filters and pcl_surface depend on VTK

filters is not dependent on VTK. Packages dependent on VTK are:
* optional: 2d, apps (most of them), io, surface
* required: apps/{cloud_composer, modeler, 3d_rec_framework}, examples, visualization

You can build PCL without VTK support using the compile flag `-DWITH_VTK:=OFF`. To test if PCL was built with VTK or not, pcl_io has a define `PCL_BUILD_WITH_VTK` (check needed for conversion to/from VTK formats)
